osu.Game/Online/RealtimeMultiplayer/StatefulMultiplayerClient.cs

```diff
@@ -0,0 +1,389 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Game.Beatmaps;
+using osu.Game.Database;
+using osu.Game.Online.API;
+using osu.Game.Online.API.Requests;
+using osu.Game.Online.Multiplayer;
+using osu.Game.Online.Multiplayer.RoomStatuses;
+using osu.Game.Rulesets;
+using osu.Game.Users;
+using osu.Game.Utils;
+
+namespace osu.Game.Online.RealtimeMultiplayer
+{
+    public abstract class StatefulMultiplayerClient : Component, IMultiplayerClient, IMultiplayerRoomServer
+    {
+        /// <summary>
+        /// Invoked when any change occurs to the multiplayer room.
+        /// </summary>
+        public event Action? RoomChanged;
+
+        /// <summary>
+        /// Invoked when the multiplayer server requests the current beatmap to be loaded into play.
+        /// </summary>
+        public event Action? LoadRequested;
+
+        /// <summary>
+        /// Invoked when the multiplayer server requests gameplay to be started.
+        /// </summary>
+        public event Action? MatchStarted;
+
+        /// <summary>
+        /// Invoked when the multiplayer server has finished collating results.
+        /// </summary>
+        public event Action? ResultsReady;
+
+        /// <summary>
+        /// Whether the <see cref="StatefulMultiplayerClient"/> is currently connected.
+        /// </summary>
+        public abstract IBindable<bool> IsConnected { get; }
+
+        /// <summary>
+        /// The joined <see cref="MultiplayerRoom"/>.
+        /// </summary>
+        public MultiplayerRoom? Room { get; private set; }
+
+        /// <summary>
+        /// The users currently in gameplay.
+        /// </summary>
+        public readonly BindableList<int> PlayingUsers = new BindableList<int>();
+
+        [Resolved]
+        private UserLookupCache userLookupCache { get; set; } = null!;
+
+        [Resolved]
+        private IAPIProvider api { get; set; } = null!;
+
+        [Resolved]
+        private RulesetStore rulesets { get; set; } = null!;
+
+        private Room? apiRoom;
+        private int playlistItemId; // Todo: THIS IS SUPER TEMPORARY!!
```

Comment:
The playlist eventually needs to be updated to a new playlist item ID, at which point it'll be returned in the `MultiplayerRoomSettings` object.

Reply:
Is fine, but probably make the comment mention why it's required/temporary just to have a record in the code itself (not sure how soon this will get fixed, needs some further server-side consideration).